### PR TITLE
Use version-specific prefs files when running Firefox

### PR DESCRIPTION
**Pull request by jgraham**
Firefox requires a prefs file to be loaded to ensure that it doesn't
make external network connections, and to make sure that other
settings are appropriate for testing. Previously we always used the
version of the prefs file from master, which is usually fine for
nightly builds, but doesn't work with release builds if it happens
that a pref changed.

This change uses the correct release version of the prefs file for
firefox releases and beta versions. It continues to use the master
version for any nightly build; this could be fixed in some cases if we
are able to access the actual commit hash used to build Firefox, but
probably isn't too bad an approximation.

The caching algorithm was changed so that release versions of the
prefs are cached forever, and the nightly version is updated once per
day (although this doesn't quite match the nightly release cadence,
it's only going to fail in edge cases where the prefs were changed in
the file, but the nightly version was not yet updated, of vice-versa.)

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/10443)

### CODE REVIEW

File: tools/wpt/browser.py [hunk 1]
```diff
@@ -179,22 +180,70 @@ def find_certutil(self):
...
+        version_re = re.compile("Mozilla Firefox (\d+\.\d+(?:\.\d+)?)(a|b)?")
+        proc = subprocess.Popen([binary, "--version"], stdout=subprocess.PIPE)
+        stdout, _ = proc.communicate()
+        stdout.strip()
+        m = version_re.match(stdout)
+        if not m:
+            return None, "nightly"
+        version, status = m.groups()
+        channel = {"a": "nightly", "b": "beta", None: "stable"}
+        return version, channel.get(status, "stable")
```
Comment (jugglinmike) [inline tools/wpt/browser.py]: I don't know which of the two approaches to defining the default value is more Pythonic, but I don't think we need them both.

File: tools/wpt/browser.py [hunk 2]
```diff
@@ -179,22 +180,70 @@ def find_certutil(self):
...
+            repo = "https://hg.mozilla.org/releases/mozilla-release"
+            tag = "FIREFOX_%s_RELEASE" % version.replace(".", "_")
+        else:
+            repo = "https://hg.mozilla.org/mozilla-central"
+            if channel == "beta":
+                tag = "FIREFOX_%s_BETA" % version.split(".", 1)[0]
+            else:
+                # Always use tip as the tag for nightly; this isn't quite right
+                # but to do better we need the actual build revision, which we
+                # can get if we have an application.ini file
```
Comment (jugglinmike) [inline tools/wpt/browser.py]: This hint for future improvement is appreciated!

File: tools/wpt/browser.py [hunk 3]
```diff
@@ -179,22 +180,70 @@ def find_certutil(self):
...
+        # If we don't have a recent download, grab the url
+        if not have_cache:
+            url = self.get_prefs_url(version, channel)
+
+            with open(cache_file, "wb") as f:
+                print("Installing test prefs from %s" % url)
+                resp = get(url)
                 f.write(resp.content)
+        else:
+            print("Using cached test prefs from %s" % cache_file)
```
Comment (jugglinmike) [inline tools/wpt/browser.py]: We have a `logger` instance in this file. Mind using that instead?

File: tools/wpt/browser.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import logging
+import glob
```
Comment (jugglinmike) [inline tools/wpt/browser.py]: This line may be removed

**Review by jugglinmike — CHANGES_REQUESTED** [review]
Looking good, James! Just a few small requests

**Comment by jgraham** [comment]
I fixed everything apart from the logger thing, which seems to be a little complicated. I need to remember what we do to make the stdlib logger work from mozlog, because at the moment it's looking pretty broken to me, but then it works later so I guess I'm missing something.

I think we should do that in a followup since it's not really a change in this PR (we were using `print` before these changes as well).

**Comment by jugglinmike** [comment]
> I think we should do that in a followup since it's not really a change in this PR (we were using print before these changes as well).

Works for me. Could you add a link here when you do?

**Review by jugglinmike — APPROVED** [review]
Tested locally with Firefox version 59.0.2 and Firefox Nightly. Prior to application of the patch, version 59.0.2 consistently fails due to its making network requests on startup. After applying the patch, these requests are correctly disabled for both versions.

Thanks, @jgraham!
